### PR TITLE
Fix coroutine scoping of Redwood Treehouse UI

### PR DESCRIPTION
**Pull request by dellisd**
Ensure that Zipline resources aren't closed until the UI's coroutine scope job completes.

This prevents a queued-up job from resuming(?) and crashing from a `ZiplineException` instead of gracefully handling a `CancellationException` when trying to collect a `ZiplineService`-backed flow.

---

- [x] `CHANGELOG.md`'s "Unreleased" section has been updated, if applicable.

### CODE REVIEW

File: redwood-treehouse-guest/src/commonMain/kotlin/app/cash/redwood/treehouse/treehouseCompose.kt [hunk 1]
```diff
@@ -121,9 +130,12 @@ private class RedwoodZiplineTreehouseUi(
   }
 
   override fun close() {
+    coroutineScope.coroutineContext.job.invokeOnCompletion {
```
Comment (swankjesse) [inline redwood-treehouse-guest/src/commonMain/kotlin/app/cash/redwood/treehouse/treehouseCompose.kt:133]: Love it

**Review by swankjesse — APPROVED** [review]

File: redwood-treehouse-guest/src/commonMain/kotlin/app/cash/redwood/treehouse/treehouseCompose.kt [hunk 2]
```diff
@@ -55,6 +59,11 @@ private class RedwoodZiplineTreehouseUi(
    */
   override val scope = (treehouseUi as? ZiplineScoped)?.scope ?: ZiplineScope()
 
+  private val coroutineScope = CoroutineScope(
+    appLifecycle.coroutineScope.coroutineContext +
+      SupervisorJob(appLifecycle.coroutineScope.coroutineContext.job),
```
Comment (JakeWharton) [inline redwood-treehouse-guest/src/commonMain/kotlin/app/cash/redwood/treehouse/treehouseCompose.kt]: Why this change? Doesn't this mean crashes in the composition won't correctly propagate?

**Comment by JakeWharton** [comment]
Also if you're fixing something please add a change log entry to accompany it.

File: redwood-treehouse-guest/src/commonMain/kotlin/app/cash/redwood/treehouse/treehouseCompose.kt (hunk 2 above)
Comment (dellisd) [reply redwood-treehouse-guest/src/commonMain/kotlin/app/cash/redwood/treehouse/treehouseCompose.kt]: As far as I can tell the unhandled exceptions still propagate up to the appLifecycle scope's job and are handled by its exception handler--but you're right, a supervisor job is probably not correct here.

We just need some per-UI scope that can be cancelled independently of the app's overall scope.

Comment (JakeWharton) [reply redwood-treehouse-guest/src/commonMain/kotlin/app/cash/redwood/treehouse/treehouseCompose.kt]: Yeah sorry the "this" was meant solely to be about the `SupervisorJob`. The subscope makes sense 👍

File: redwood-treehouse-guest/src/commonMain/kotlin/app/cash/redwood/treehouse/treehouseCompose.kt [hunk 3]
```diff
@@ -55,6 +58,11 @@ private class RedwoodZiplineTreehouseUi(
    */
   override val scope = (treehouseUi as? ZiplineScoped)?.scope ?: ZiplineScope()
 
+  private val coroutineScope = CoroutineScope(
+    appLifecycle.coroutineScope.coroutineContext +
+      appLifecycle.coroutineScope.coroutineContext.job,
+  )
```
Comment (JakeWharton) [inline redwood-treehouse-guest/src/commonMain/kotlin/app/cash/redwood/treehouse/treehouseCompose.kt]:
```suggestion
  private val coroutineScope = CoroutineScope(
    appLifecycle.coroutineScope.coroutineContext +
      Job(appLifecycle.coroutineScope.coroutineContext.job),
  )
```
I think you probably still want to put in a new `Job` which gets canceled. Or does creating a new `CoroutineScope` do that automatically? If so, you probably don't need the `+ ...` part.

Comment (dellisd) [reply redwood-treehouse-guest/src/commonMain/kotlin/app/cash/redwood/treehouse/treehouseCompose.kt]: Oh woops, that's right. A new Job is only created if there's no Job in the context. This scope's job should be a child of the parent scope's job, but this is giving both scopes the same job. Will fix!
